Article route / controller refactor

trainerbill opened the pull request:
This has been bothering me for awhile.  I refactored the Article controller and routes to be more in line with ui-router.

1.  Separated controllers into specific CRUD actions
2.  Removed ng controller from views
3.  Added controllers to routes
4.  Prevent loading of article every time route loads if the article is the same id that is already in scope.article

Benefits
1.  More inline with ui-router standards
2.  Less API calls since we are checking to see if the right article is already added.
3.  Ability to check authentication in EditController

Obviously the tests are going to fail.  I will fix them once I get buy in and the code is set.

@lirantal @ilanbiala @rhutchison @mleanos @codydaig

trainerbill commented:
Anyone have any comments?

rhutchison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,10 +1,58 @@
+  function ($scope, $state, $stateParams, Articles) {
+    //Check if article is loaded if not load it
+    if ($scope.article === undefined || typeof $scope.article !== 'object' || $scope.article._id === undefined || $scope.article._id !== $stateParams.articleId) {
+      $scope.loadArticleById($stateParams.articleId);
```
Article should come from route resolve, not from controller

rhutchison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -1,10 +1,58 @@
+
+    //Check if article is loaded if not load it
+    if ($scope.article === undefined || typeof $scope.article !== 'object' || $scope.article._id === undefined || $scope.article._id !== $stateParams.articleId) {
+      $scope.loadArticleById($stateParams.articleId);
```
Article should come from route resolve, not from controller.

rhutchison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -24,45 +72,35 @@ angular.module('articles').controller('ArticlesController', ['$scope', '$statePa
-    $scope.findOne = function () {
-      $scope.article = Articles.get({
-        articleId: $stateParams.articleId
+    $scope.authCheck = function () {
```
This should go in the route resolve

rhutchison commented:
I think it can be simplified with 2 controllers - ArticleListController, ArticleController.  Controllers should be separate files.

mleanos commented:
+1 on using just two controllers, and as separate files. 

@trainerbill There are a lot of changes revolving around the Articles module, right now. I have an open PR #807  that contradicts this one in a sense. I'm not even sure if my proposed changes are what we want to go with. But I wanted to make sure you were aware of it.

trainerbill commented:
Thanks @mleanos.  I think that the 2 PR's can coexist.  Just keep yours to only admin functionality.  I really think the article module needs to be refactored separately.

trainerbill commented:
Alright gents.  I moved the load and authcheck to the $state resolve.  Split out into 2 controllers.  The only thing I don't like is having a mock article resolve for the create route since you have to resolve article to use the ArticlesController.

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,81 @@ angular.module('articles').config(['$stateProvider',
+        controller: 'ArticlesController',
+        resolve: {
+          article: function (Articles, $stateParams, $state) {
+            return Articles.get({
```
just return Articles.get(..);

the 404 should be handled by the http interceptor (https://github.com/meanjs/mean/pull/796/files#diff-90593c1684669b4a0cc185ec424f0a24R26)

trainerbill replied:
I agree... but I doubt they will merge this in without the interceptor and I don't know if they like the interceptors.... So we would need both PR's to merge together.  Which I am alright with.

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,81 @@ angular.module('articles').config(['$stateProvider',
+            }).$promise.then(
+              function (article) {
+                //Auth Check
+                if (article.user._id !== Authentication.user._id) {
```
You still do not check for admin

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,81 @@ angular.module('articles').config(['$stateProvider',
+                }
+                return article;
+              },
+              function () {
```
This would be handled by http interceptor, not required here

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,81 @@ angular.module('articles').config(['$stateProvider',
+        controller: 'ArticlesListController',
+        resolve:{
+          articles: function (Articles, $stateParams, $state) {
+            return Articles.query().$promise.then(
```
`return Articles.query();`

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,81 @@ angular.module('articles').config(['$stateProvider',
+        },
+        resolve: {
+          article: function () {
+            return { mock: 'mock' };
```
just `return {};`

trainerbill commented:
Can I get a couple LGTM?  I want to fix the tests but need to settle on the code first.

codydaig commented:
LGTM. Haven't tested, but like and agree with the concept.

simison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,57 @@ angular.module('articles').config(['$stateProvider',
+                return article;
+              });
+          }
+        },
```
Wait, how can the backend return something you're not authorised to see? You should get 403 in that case.

In any case, this is a lot of logic for route config. Thoughts?

trainerbill replied:
Because we are not hitting the backend until the actual update of the article.  So the article is loaded in scope when it is viewed.  Now when you put /edit in the URL you will get to the edit route because we are not hitting the backend.  If you tried to update you would get authorization error but that does not prevent you from getting to the route.  This will.

simison replied:
ic, makes sense now. :thumbsup:

simison commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -25,44 +25,29 @@ angular.module('articles').controller('ArticlesController', ['$scope', '$statePa
+    $scope.remove = function () {
+      if (confirm('Are you sure you want to delete?')) {
+        $scope.article.$remove(function (article) {
+          if ($scope.articles !== undefined && $scope.articles.length > 0) {
```
PS. https://docs.angularjs.org/api/ng/function/angular.isUndefined :-)

rhutchison commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,57 @@ angular.module('articles').config(['$stateProvider',
+        templateUrl: 'modules/articles/client/views/list-articles.client.view.html',
+        controller: 'ArticlesListController',
+        resolve:{
+          articles: function (Articles, $stateParams, $state) {
```
don't like having the articles list in the resolve, the service should still be a part of the controller.

trainerbill replied:
Why not?  Do we not want the most recent data on the route resolve?  Not sure why we would want to call it in the controller after moving everything else to resolve.

mleanos replied:
In my opinion, `resolve`'s on a route are better suited for when you're interacting with an instance of a resource/entity. When a view is dependent on specific data actually existing at the time of loading. In the case of lists, it's not critical that there's data when the view loads or even if such data exists.

The concern of the route config (`resolve` in this case) should be to make sure the view has everything it needs in order to function properly, and for authorization purposes.

ilanbiala commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,52 @@ angular.module('articles').config(['$stateProvider',
+        },
+        resolve: {
+          article: function () {
+            return { };
```
no space, but on another note, why do we need this?

trainerbill replied:
Because article controller is now expecting to resolve an article.  Its either send a blank object or split it to another controller

ilanbiala replied:
This seems too hacky, I'll try to think of a better way.

ilanbiala commented on modules/articles/client/config/articles.client.routes.js:
```diff
@@ -12,22 +12,52 @@ angular.module('articles').config(['$stateProvider',
+            }).$promise.then(function (article) {
+                //Auth Check
+                if (article.user._id !== Authentication.user._id && Authentication.user.roles.indexOf('admin') === -1) {
+                  //TODO  Change to unauthorized when that PR is merged
```
What about this TODO?

trainerbill replied:
I will change it now that the unauthorized PR is merged

ilanbiala commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -63,22 +48,25 @@ angular.module('articles').controller('ArticlesController', ['$scope', '$statePa
 
       article.$update(function () {
-        $location.path('articles/' + article._id);
+        $state.go('articles.view',{articleId: article._id});
```
comma spacing

ilanbiala commented on modules/articles/client/controllers/articles.client.controller.js:
```diff
@@ -63,22 +48,25 @@ angular.module('articles').controller('ArticlesController', ['$scope', '$statePa
-        articleId: $stateParams.articleId
-      });
-    };
+    //Not sure if this is needed anymore?
```
Is this still needed?

trainerbill replied:
Yeah its needed to display edit/remove on the view article.  Would you all be opposed to do a scope variable canEdit canRemove instead of putting the whole Authentication object in scope and checking in the view?

ilanbiala commented on modules/articles/client/controllers/articles.client.list.controller.js:
```diff
@@ -0,0 +1,12 @@
+.controller('ArticlesListController', ['$scope', 'Articles',
+  function ($scope, Articles) {
+    //Query Articles
+    Articles.query().$promise.then(function (data) {
```
call data articles, no?

trainerbill replied:
changed.  I just always use data.

ilanbiala commented on modules/articles/tests/client/articles.client.controller.tests.js:
```diff
@@ -62,149 +62,43 @@
       // Initialize the Articles controller.
-      ArticlesController = $controller('ArticlesController', {
+      ArticlesListController = $controller('ArticlesListController', {
         $scope: scope
```
Why aren't you injecting Articles as well?

trainerbill replied:
I had the wrong file names for the article and article.list tests

ilanbiala commented on modules/articles/tests/client/articles.client.controller.tests.js:
```diff
@@ -62,149 +62,43 @@
 
         // Test URL redirection after the article was created
-        expect($location.path.calls.mostRecent().args[0]).toBe('articles/' + mockArticle._id);
+        //expect($state.go).toHaveBeenCalledWith('articles.view', {articleId: mockArticle._id});
```
Why was this commented out?

trainerbill replied:
Because we don't redirect on article list.  I removed the comments

trainerbill commented:
@ilanbiala This PR is really messed up.  Evidently I merged master in a couple times.  You care if I close and reopen with just the changes?  This commit is killing me on rebase:

https://github.com/trainerbill/mean/commit/21fa7dc1ead3777fbfa8bba68067b91af8b46e71

Looks like I did something pretty dumb on it.

ilanbiala commented:
Please do whatever you need to keep it tidy.

trainerbill commented:
Closing in favor of https://github.com/meanjs/mean/pull/874

mleanos commented on q:
```diff
@@ -0,0 +1,2016 @@
+[33mcommit 4f89e8266e5b2b5f6b1836097436a79a4162bda1[m
```
@trainerbill What is this file?
